#[derive(Clone)] fails for HRTB function type taking an associated type

The following code does not compile, even though it probably should:
```rust
trait SomeTrait {
    type SomeType<'a>;
}

#[derive(Clone)]
struct Foo<T: SomeTrait> {
    x: for<'a> fn(T::SomeType<'a>)
}
```

([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=862d39bfcbaeb7d1e462ee5408c37585))

The error message on rust 1.76.0 stable is:
<details>

```
   Compiling playground v0.0.1 (/playground)
error[E0261]: use of undeclared lifetime name `'a`
 --> src/lib.rs:7:31
  |
7 |     x: for<'a> fn(T::SomeType<'a>)
  |                               ^^ undeclared lifetime
  |
  = note: for more information on higher-ranked polymorphism, visit https://doc.rust-lang.org/nomicon/hrtb.html
help: consider making the bound lifetime-generic with a new `'a` lifetime
  |
5 | #[derive(for<'a> Clone)]
  |          +++++++
help: consider introducing lifetime `'a` here
  |
6 | struct Foo<'a, T: SomeTrait> {
  |            +++

For more information about this error, try `rustc --explain E0261`.
error: could not compile `playground` (lib) due to 1 previous error
```

</details>

(The note being incorrect is a separate bug, already filed as #107694.)

The error message on rust `1.78.0-nightly (2024-03-16 766bdce744d531267d53)` is:

<details>

```
   Compiling playground v0.0.1 (/playground)
error[E0261]: use of undeclared lifetime name `'a`
 --> src/lib.rs:7:31
  |
6 | struct Foo<T: SomeTrait> {
  |            - help: consider introducing lifetime `'a` here: `'a,`
7 |     x: for<'a> fn(T::SomeType<'a>)
  |                               ^^ undeclared lifetime
  |
  = note: for more information on higher-ranked polymorphism, visit https://doc.rust-lang.org/nomicon/hrtb.html

For more information about this error, try `rustc --explain E0261`.
error: could not compile `playground` (lib) due to 1 previous error
```

</details>

Note that manually implementing a Clone compiles fine without any errors:
```rust
trait SomeTrait {
    type SomeType<'a>;
}

struct Foo<T: SomeTrait> {
    x: for<'a> fn(T::SomeType<'a>)
}

impl<T: SomeTrait> Clone for Foo<T> {
    fn clone(&self) -> Self {
        Self {
            x: self.x
        }
    }
}
```
([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d274800c8401926b9773910ece53684d))

--- a/compiler/rustc_builtin_macros/src/deriving/generic/mod.rs
+++ b/compiler/rustc_builtin_macros/src/deriving/generic/mod.rs
@@ -412,6 +412,15 @@ fn find_type_parameters(
 
     impl<'a, 'b> visit::Visitor<'a> for Visitor<'a, 'b> {
         fn visit_ty(&mut self, ty: &'a ast::Ty) {
+            let stack_len = self.bound_generic_params_stack.len();
+            if let ast::TyKind::BareFn(bare_fn) = &ty.kind
+                && !bare_fn.generic_params.is_empty()
+            {
+                // Given a field `x: for<'a> fn(T::SomeType<'a>)`, we wan't to account for `'a` so
+                // that we generate `where for<'a> T::SomeType<'a>: ::core::clone::Clone`. #122622
+                self.bound_generic_params_stack.extend(bare_fn.generic_params.iter().cloned());
+            }
+
             if let ast::TyKind::Path(_, path) = &ty.kind
                 && let Some(segment) = path.segments.first()
                 && self.ty_param_names.contains(&segment.ident.name)
@@ -422,7 +431,8 @@ fn find_type_parameters(
                 });
             }
 
-            visit::walk_ty(self, ty)
+            visit::walk_ty(self, ty);
+            self.bound_generic_params_stack.truncate(stack_len);
         }
 
         // Place bound generic params on a stack, to extract them when a type is encountered.

--- a/tests/ui/derives/derive-hrtb-for-bare-fn-field-with-lifetime.rs
+++ b/tests/ui/derives/derive-hrtb-for-bare-fn-field-with-lifetime.rs
@@ -1,0 +1,13 @@
+//@ run-pass
+// Issue #122622: `#[derive(Clone)]` should work for HRTB function type taking an associated type
+#![allow(dead_code)]
+trait SomeTrait {
+    type SomeType<'a>;
+}
+
+#[derive(Clone)]
+struct Foo<T: SomeTrait> {
+    x: for<'a> fn(T::SomeType<'a>)
+}
+
+fn main() {}

